### PR TITLE
More general LAMMPS code design info for the Programmer guide section of the manual

### PR DESCRIPTION
**Pull request by akohlmey**
**Summary**

This PR provides some information on the general code design of LAMMPS and specifically discusses some aspects of C++ syntax and semantics in a way that is (hopefully) accessible to developers with limited C++ experience.
Also some more recommendations and explanations on which code constructs to use and why.

**Author(s)**

Axel Kohlmeyer, Temple U

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system

### CODE REVIEW

File: doc/src/Developer_cxx_vs_c_style.rst [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+===========================
+
+LAMMPS makes extensive use of the object oriented programming (OOP)
+principles of *compositing* and *inheritance*. Classes like the
+``LAMMPS`` class are a **composite** containing pointers to instances of
+other classes like ``Atom``, ``Comm``, ``Force``, ``Neighbor``,
+``Modify``, and so on. Each of these classes implement certain
+functionality by storing and manipulating data related to the simulation
+and providing member functions that trigger certain actions.  Some of
+those classes like ``Force`` and a composite again containing instances
```
Comment (athomps) [inline doc/src/Developer_cxx_vs_c_style.rst:83]: >>> are not and

File: doc/src/Developer_cxx_vs_c_style.rst [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+- a failing "new" will trigger an exception which is also captured by LAMMPS and
+  triggers a global abort
+- allocation of multi-dimensional arrays will be done in a C compatible fashion
+  but so that the storage of the actual data is stored in one large consecutive block
+  and thus when MPI communication is needed, only this storage needs to be
+  communicated (similar to Fortran arrays)
+- the "destroy()" and "sfree()" functions may safely be called on NULL pointers
+- the "destroy()" functions will nullify the pointer variables making
+  "use after free" errors easy to detect
+- it is possible to use a large than default memory alignment (not on all operating
```
Comment (athomps) [inline doc/src/Developer_cxx_vs_c_style.rst:362]: large -> larger

**Review by athomps — APPROVED** [review]
Very nice. I approve, but added comments on several minor typographical errors.

File: doc/src/Developer_cxx_vs_c_style.rst [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+class variant was instantiated.  In the LAMMPS documentation those
+derived classes are usually referred to a "styles", e.g.  pair styles,
+fix styles, atom styles and so on.
+
+This is the origin of the flexibility of LAMMPS and facilitates for
+example to compute forces for very different non-bonded potential
+functions by having different pair styles (implemented as different
+classes derived from the ``Pair`` class) where the evaluation of the
+potential function is confined to the implementation of the individual
+classes.  Whenever a new :doc:`pair_style` or :doc:`bond_style` or
```
[the remark this answers is not in the document]
Comment (akohlmey) [reply doc/src/Developer_cxx_vs_c_style.rst:106]: This *was* mentioned a couple paragraphs above (around line 80), but I have made it more explicit and named a few examples.

**Pull request merged**